sw/device/tests/rv_core_ibex_rnd_test.c

```diff
@@ -0,0 +1,99 @@
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/base/memory.h"
+#include "sw/device/lib/base/mmio.h"
+#include "sw/device/lib/dif/dif_edn.h"
+#include "sw/device/lib/dif/dif_rv_core_ibex.h"
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/testing/entropy_testutils.h"
+#include "sw/device/lib/testing/test_framework/check.h"
+#include "sw/device/lib/testing/test_framework/ottf_main.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+#include "rv_core_ibex_regs.h"
+
+OTTF_DEFINE_TEST_CONFIG();
+
+dif_result_t rv_core_ibex_read_rnd_data_followed_by_status(
+    const dif_rv_core_ibex_t *rv_core_ibex, uint32_t *data,
+    dif_rv_core_ibex_rnd_status_t *followed_status) {
+  if (rv_core_ibex == NULL || data == NULL || followed_status == NULL) {
+    return kDifBadArg;
+  }
+  uint32_t reg = mmio_region_read32(rv_core_ibex->base_addr,
+                                    RV_CORE_IBEX_RND_STATUS_REG_OFFSET);
+  if (!bitfield_bit32_read(reg, RV_CORE_IBEX_RND_STATUS_RND_DATA_VALID_BIT)) {
+    return kDifError;
+  }
+
+  *data = mmio_region_read32(rv_core_ibex->base_addr,
+                             RV_CORE_IBEX_RND_DATA_REG_OFFSET);
+
+  // There are only 10 cycles between reading RND and status returning back
+  // high.
+  *followed_status = mmio_region_read32(rv_core_ibex->base_addr,
+                                        RV_CORE_IBEX_RND_STATUS_REG_OFFSET);
+  return kDifOk;
+}
+
+dif_result_t rv_core_ibex_multiple_unchecked_read_rnd_data(
+    const dif_rv_core_ibex_t *rv_core_ibex, const uint32_t repetitions,
+    uint32_t *data) {
+  if (rv_core_ibex == NULL || data == NULL) {
+    return kDifBadArg;
+  }
+
+  for (uint32_t i = 0; i < repetitions; i++) {
+    *data = mmio_region_read32(rv_core_ibex->base_addr,
+                               RV_CORE_IBEX_RND_DATA_REG_OFFSET);
+  }
+  return kDifOk;
+}
+
+#define RANDOM_DATA_READS 16
+
+bool test_main(void) {
+  // Verify the functionality of the random number generation CSRs.
+
+  // Add variables to the stack.
+  dif_rv_core_ibex_t rv_core_ibex;
+  uint32_t rnd_data, previous_rnd_data;
+  dif_rv_core_ibex_rnd_status_t rnd_status = 0;
+  const uint32_t kUncheckedReadRepititions = 64;
+
+  // Enable entropy complex, CSRNG and EDN so Ibex can get entropy.
+  entropy_testutils_boot_mode_init();
+
+  // Initialize Ibex.
+  CHECK_DIF_OK(dif_rv_core_ibex_init(
+      mmio_region_from_addr(TOP_EARLGREY_RV_CORE_IBEX_CFG_BASE_ADDR),
+      &rv_core_ibex));
+
+  // Perform multiple reads from `RND_DATA` polling `RND_STATUS` in between to
+  // only read valid data. Check different random bits are provided each time.
+  for (int i = 0; i < RANDOM_DATA_READS; i++) {
+    while (rnd_status == 0) {
+      CHECK_DIF_OK(dif_rv_core_ibex_get_rnd_status(&rv_core_ibex, &rnd_status));
+    }
+    CHECK_DIF_OK(dif_rv_core_ibex_read_rnd_data(&rv_core_ibex, &rnd_data));
+    if (i != 0) {
+      CHECK(rnd_data != previous_rnd_data);
+    }
```

Comment:
I don't think entropy_src will ever return 0s, so this `if` clause can be omitted. In fact, we should probably check `rnd_data` is never all 0s or all 1s every iteration.